model_compression_toolkit/core/common/model_collector.py

```diff
@@ -154,14 +184,35 @@ def infer(self, inputs_list: List[np.ndarray]):
 
         # TODO: Thinking about delegating collections to framework
         # TODO: migrate datasets to framework datasets
-        tensor_data = self.fw_impl.run_model_inference(self.model, inputs_list)
-        for td, sc in zip(tensor_data, self.stats_containers_list):
+        compute_hessians = self.hessian_service is not None and self.qc.activation_error_method == QuantizationErrorMethod.HMSE
+        tensor_data = self.fw_impl.run_model_inference(self.model, inputs_list, requires_grad=compute_hessians)
+
+        if compute_hessians:
+            request = HessianScoresRequest(
+                mode=HessianMode.ACTIVATION,
+                granularity=HessianScoresGranularity.PER_ELEMENT,
+                target_nodes=self.intermediate_output_tensors,
+                data_loader=None,
+                n_samples=inputs_list[0].shape[0],
+                compute_from_tensors=True
+            )
+            hessian_data = self.hessian_service.fetch_hessian(request=request,
+                                                              activation_tensors=tensor_data)
+            hessian_data = list(hessian_data.values())
+        else:
+            hessian_data = []
+
+        # Hessian is not calculated for the output, add "None" as weights for output tenosrs
+        hessian_data += [None for _ in range(len(tensor_data) - len(hessian_data))]
```

Comment:
If this is done to overcome the outputs issue, I think it's better to simply use the graph number of outputs instead of concluding it from this.